pkg/trainer/training.go

```diff
@@ -447,118 +449,12 @@ func (j *TrainingJob) run(config *spec.ControllerConfig, stopC <-chan struct{})
 				// Return from run because we want to stop reconciling the object.
 				return
 			}
-
 		case <-time.After(reconcileInterval):
-			// TODO(jlewi): Can we determine from the TPR status whether we should
-			// Create the resources or not? We need to ensure the resources exist so for
-			// now we always call Create.
-			if j.job.Status.Phase == spec.TfJobPhaseRunning {
-				// We call Create to make sure all the resources exist and are running.
-				if cErr := j.createResources(config); cErr != nil {
-					log.Errorf("trainingJobCreateReplicas() error; %v", cErr)
-				}
-
-				state, replicaStatuses, err := j.GetStatus()
-
-				j.status.ReplicaStatuses = replicaStatuses
-				if err != nil {
-					log.Errorf("GetStatus() for job %v returned error: %v", j.job.Metadata.Name, err)
-				}
-				// TODO(jlewi): We should update the Phase if we detect the job is done.
-				if state == spec.StateFailed {
-					log.Errorf("Master failed Job: %v.", j.job.Metadata.Name)
-					j.status.SetPhase(spec.TfJobPhaseDone)
-					j.status.SetState(spec.StateFailed)
-				} else if state == spec.StateSucceeded {
-					log.Infof("Master succeeded Job: %v.", j.job.Metadata.Name)
-					j.status.SetPhase(spec.TfJobPhaseDone)
-					j.status.SetState(spec.StateSucceeded)
-				} else {
-					log.V(1).Infof("Job %v status=%v", j.job.Metadata.Name, util.Pformat(j.status))
-				}
-			}
-
-			// If the phase changed we should update the TPR.
-			if err := j.updateTPRStatus(); err != nil {
-				log.Warningf("Job %v, failed to update TPR status error: %v", j.job.Metadata.Name, err)
-			}
-
-			if j.job.Status.Phase == spec.TfJobPhaseCleanUp {
-				if cErr := j.deleteResources(); cErr != nil {
-					log.Errorf("Job %v trainingJob.Delete() error; %v", j.job.Metadata.Name, cErr)
-				}
-				// j.status.SetPhase(spec.TfJobPhaseDone)
-				// Return from run because we want to stop reconciling the object.
-				return
-			}
-
-			if rerr != nil {
-				log.Errorf("failed to reconcile job %v, error: %v", j.job.Metadata.Name, rerr)
-				break
-			}
-
-			// updateTPRStatus will update the status of the TPR with c.Status if c.Status
-			// doesn't match c.Cluster.status. So you can change c.Status in order to propogate
-			// changes to the TPR status.
-			if err := j.updateTPRStatus(); err != nil {
-				log.Warningf("Job %v; failed to update TPR status error: %v", j.job.Metadata.Name, err)
-			}
+			j.reconcile(config)
```

Comment:
seems we will do this every 8 minute intervals, if i understand this code, maybe we only need do `setup` when my job is created and phase is none.

Reply:
The reconcile function is the code that periodically checks the job and figures out what needs to be done. The reconcile function takes advantage of state (e.g. phase) preserved in the CRD to avoid repeating unnecessary steps like setup.